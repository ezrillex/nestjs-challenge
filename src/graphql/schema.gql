# ------------------------------------------------------
# THIS FILE WAS AUTOMATICALLY GENERATED (DO NOT MODIFY)
# ------------------------------------------------------

type Categories {
  id: String!
  name: String!
}

type GetCategoriesResponse {
  id: String!
  name: String!
  products: [Products!]!
}

type Images {
  id: String!
  url: String!
  created_at: DateTime!
  product_variation_id: String
}

"""
A date-time string at UTC, such as 2019-12-03T09:54:33Z, compliant with the date-time format.
"""
scalar DateTime

type ProductVariations {
  id: String!
  title: String!
  price: Float!
  stock: Int!
  images: [Images!]
}

type Products {
  id: ID!
  is_published: Boolean!
  is_deleted: Boolean!
  name: String!
  description: String!
  variations: [ProductVariations!]!
  categories: [Categories!]!
  created_by: String!
  created_at: DateTime!
  last_updated_by: String
  last_updated_at: DateTime
}

type LikesOfProducts {
  user_id: ID!
  product_variation_id: ID!
  state: Boolean!
}

type Users {
  id: ID!
  first_name: String!
  last_name: String!
  email: String!
  role: String!
  created_at: DateTime!
  login_at: DateTime!
  password_last_updated: DateTime!
  likes_products: [LikesOfProducts!]!
}

type CartItems {
  id: ID!
  cart_owner: Users!
  product_variation: ProductVariations!
  quantity: Int!
}

type OrderItems {
  id: ID!
  product_variation: ProductVariations
  quantity: Int!
  price_purchased_at: Float!
}

type PaymentIntents {
  id: ID!
  status: String
  created_at: DateTime!
  stripe_event_id: ID!
}

type Orders {
  id: ID!
  user: Users!
  payments: [PaymentIntents!]!
  paymentStatus: String!
  orderStatus: String!
  order_items: [OrderItems!]!
}

type Query {
  user(id: ID!): Users!
  getLikes: [ProductVariations!]
  getProducts(GetProductsInput: GetProductsInput!): [Products!]
  getProductById(id: String!): Products
  getCategories: [GetCategoriesResponse!]
  getCartItems: [CartItems!]
  getOrders(GetOrdersInput: GetOrdersInput!): [Orders!]
  getOrder(order_id: String!): Orders
}

input GetProductsInput {
  first: Int
  offset: Int
  categoryFilter: [ID!]
  search: String
  likedOnly: Boolean
  omitImages: Boolean
  sort: String
}

input GetOrdersInput {
  first: Int
  offset: Int
  client_id: ID
}

type Mutation {
  toggleLike(product_variation_id: String!): LikesOfProducts
  createProduct(CreateProductInput: CreateProductInput!): Products
  updateProduct(UpdateProductInput: UpdateProductInput!): Products
  deleteProduct(product_id: String!): String
  updateProductVariation(UpdateProductVariationInput: UpdateProductVariationInput!): ProductVariations
  createProductVariation(CreateProductVariationInput: CreateProductVariationInput!): ProductVariations
  deleteProductVariation(variation_id: String!): String
  createCategory(name: String!): Categories
  deleteCategory(id: ID!): String
  addToCart(variation_id: ID!, quantity: Int!): CartItems
  removeFromCart(product_variation_id: String!): String
  createOrder: Orders
}

input CreateProductInput {
  name: String!
  description: String!
  categories: [ID!]
  variations: [CreateVariationInput!]!
}

input CreateVariationInput {
  title: String!
  price: Float!
  stock: Float!
}

input UpdateProductInput {
  id: ID!
  name: String
  description: String
  categories: [ID!]
  is_published: Boolean
}

input UpdateProductVariationInput {
  id: ID!
  title: String
  price: Float
  stock: Int
}

input CreateProductVariationInput {
  product_id: ID!
  title: String!
  price: Float!
  stock: Int!
}