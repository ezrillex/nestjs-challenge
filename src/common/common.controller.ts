import { Controller, Get, HttpCode, HttpStatus } from '@nestjs/common';
import { Public } from './decorators/public.decorator';

@Controller('common')
export class CommonController {
  @HttpCode(HttpStatus.NO_CONTENT)
  @Public()
  @Get()
  async healthService(): Promise<void> {}
}
