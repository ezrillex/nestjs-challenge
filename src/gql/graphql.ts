
/*
 * -------------------------------------------------------
 * THIS FILE WAS AUTOMATICALLY GENERATED (DO NOT MODIFY)
 * -------------------------------------------------------
 */

/* tslint:disable */
/* eslint-disable */

export class GetProductsInput {
    first?: Nullable<number>;
    offset?: Nullable<number>;
    categoryFilter?: Nullable<string[]>;
    search?: Nullable<string>;
    likedOnly?: Nullable<boolean>;
    omitImages?: Nullable<boolean>;
    sort?: Nullable<string>;
}

export class CreateProductInput {
    name: string;
    description: string;
    categories?: Nullable<string[]>;
    variations: CreateVariationInput[];
}

export class CreateVariationInput {
    title: string;
    price: number;
    stock: number;
}

export class UpdateProductInput {
    id: string;
    name?: Nullable<string>;
    description?: Nullable<string>;
    categories?: Nullable<string[]>;
    is_published?: Nullable<boolean>;
}

export class UpdateProductVariationInput {
    id: string;
    title?: Nullable<string>;
    price?: Nullable<number>;
    stock?: Nullable<number>;
}

export class CreateProductVariationInput {
    product_id: string;
    title: string;
    price: number;
    stock: number;
}

export class CreateCategoryInput {
    name: string;
}

export class Images {
    __typename?: 'Images';
    id: string;
    url: string;
    created_at: string;
    product_variation_id?: Nullable<string>;
}

export class ProductVariations {
    __typename?: 'ProductVariations';
    id: string;
    title: string;
    price: number;
    stock: number;
    images?: Nullable<Images[]>;
}

export class Categories {
    __typename?: 'Categories';
    id: string;
    name: string;
    products: Products[];
}

export class Products {
    __typename?: 'Products';
    id: string;
    is_published: boolean;
    is_deleted: boolean;
    name: string;
    description: string;
    likes: ProductVariations[];
    variations: ProductVariations[];
    categories: Categories[];
    created_by: string;
    created_at: DateTime;
    last_updated_by?: Nullable<string>;
    last_updated_at?: Nullable<string>;
}

export class Users {
    __typename?: 'Users';
    id: string;
    first_name: string;
    last_name: string;
    email: string;
    likes_products: LikesOfProducts[];
}

export class LikesOfProducts {
    __typename?: 'LikesOfProducts';
    id: string;
    liked_by: Users;
    likes_product_variation: ProductVariations;
}

export class CartItems {
    __typename?: 'CartItems';
    id: string;
    cart_owner: Users;
    product_variation: ProductVariations;
    quantity: number;
}

export class OrderItems {
    __typename?: 'OrderItems';
    id: string;
    product_variation: ProductVariations;
    quantity: number;
    price_purchased_at: number;
}

export class Orders {
    __typename?: 'Orders';
    id: string;
    user: Users;
    paymentStatus: string;
    orderStatus: string;
    order_items: OrderItems[];
}

export abstract class IQuery {
    __typename?: 'IQuery';

    abstract getProducts(GetProductsInput: GetProductsInput): Nullable<Products[]> | Promise<Nullable<Products[]>>;

    abstract getProductById(id: string): Nullable<Products> | Promise<Nullable<Products>>;

    abstract getLikes(): Nullable<LikesOfProducts[]> | Promise<Nullable<LikesOfProducts[]>>;

    abstract getCartItems(): Nullable<CartItems[]> | Promise<Nullable<CartItems[]>>;

    abstract getOrders(): Nullable<Orders[]> | Promise<Nullable<Orders[]>>;

    abstract getOrder(order_id: string): Nullable<Orders> | Promise<Nullable<Orders>>;

    abstract getClientOrders(client_id?: Nullable<string>): Nullable<Orders[]> | Promise<Nullable<Orders[]>>;
}

export abstract class IMutation {
    __typename?: 'IMutation';

    abstract createProduct(CreateProductInput: CreateProductInput): Nullable<string> | Promise<Nullable<string>>;

    abstract updateProduct(UpdateProductInput: UpdateProductInput): Nullable<string> | Promise<Nullable<string>>;

    abstract updateProductVariation(UpdateProductVariationInput: UpdateProductVariationInput): Nullable<string> | Promise<Nullable<string>>;

    abstract createProductVariation(CreateProductVariationInput: CreateProductVariationInput): Nullable<string> | Promise<Nullable<string>>;

    abstract deleteProductVariation(variation_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract deleteProduct(product_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract createCategory(CreateCategoryInput: CreateCategoryInput): Nullable<string> | Promise<Nullable<string>>;

    abstract deleteCategory(id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract likeProduct(variation_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract removeLikeProduct(like_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract addToCart(variation_id: string, quantity: number): Nullable<string> | Promise<Nullable<string>>;

    abstract removeFromCart(cart_item_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract createOrder(): Nullable<string> | Promise<Nullable<string>>;
}

export type DateTime = any;
type Nullable<T> = T | null;
