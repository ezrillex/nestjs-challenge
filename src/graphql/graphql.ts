
/*
 * -------------------------------------------------------
 * THIS FILE WAS AUTOMATICALLY GENERATED (DO NOT MODIFY)
 * -------------------------------------------------------
 */

/* tslint:disable */
/* eslint-disable */

export class GetProductsInput {
    first?: Nullable<number>;
    offset?: Nullable<number>;
    categoryFilter?: Nullable<string[]>;
    search?: Nullable<string>;
    likedOnly?: Nullable<boolean>;
    omitImages?: Nullable<boolean>;
    sort?: Nullable<string>;
}

export class GetOrdersInput {
    first?: Nullable<number>;
    offset?: Nullable<number>;
    client_id?: Nullable<string>;
}

export class CreateProductInput {
    name: string;
    description: string;
    categories?: Nullable<string[]>;
    variations: CreateVariationInput[];
}

export class CreateVariationInput {
    title: string;
    price: number;
    stock: number;
}

export class UpdateProductInput {
    id: string;
    name?: Nullable<string>;
    description?: Nullable<string>;
    categories?: Nullable<string[]>;
    is_published?: Nullable<boolean>;
}

export class UpdateProductVariationInput {
    id: string;
    title?: Nullable<string>;
    price?: Nullable<number>;
    stock?: Nullable<number>;
}

export class CreateProductVariationInput {
    product_id: string;
    title: string;
    price: number;
    stock: number;
}

export class Categories {
    __typename?: 'Categories';
    id: string;
    name: string;
}

export class GetCategoriesResponse {
    __typename?: 'GetCategoriesResponse';
    id: string;
    name: string;
    products: Products[];
}

export class Images {
    __typename?: 'Images';
    id: string;
    url: string;
    created_at: DateTime;
    product_variation_id?: Nullable<string>;
}

export class ProductVariations {
    __typename?: 'ProductVariations';
    id: string;
    title: string;
    price: number;
    stock: number;
    images?: Nullable<Images[]>;
}

export class Products {
    __typename?: 'Products';
    id: string;
    is_published: boolean;
    is_deleted: boolean;
    name: string;
    description: string;
    variations: ProductVariations[];
    categories: Categories[];
    created_by: string;
    created_at: DateTime;
    last_updated_by?: Nullable<string>;
    last_updated_at?: Nullable<DateTime>;
}

export class LikesOfProducts {
    __typename?: 'LikesOfProducts';
    user_id: string;
    product_variation_id: string;
    state: boolean;
}

export class Users {
    __typename?: 'Users';
    id: string;
    first_name: string;
    last_name: string;
    email: string;
    role: string;
    created_at: DateTime;
    login_at: DateTime;
    password_last_updated: DateTime;
    likes_products: LikesOfProducts[];
}

export class CartItems {
    __typename?: 'CartItems';
    id: string;
    cart_owner: Users;
    product_variation: ProductVariations;
    quantity: number;
}

export class OrderItems {
    __typename?: 'OrderItems';
    id: string;
    product_variation: ProductVariations;
    quantity: number;
    price_purchased_at: number;
}

export class PaymentIntents {
    __typename?: 'PaymentIntents';
    id: string;
    status?: Nullable<string>;
    created_at: DateTime;
    stripe_event_id: string;
}

export class Orders {
    __typename?: 'Orders';
    id: string;
    user: Users;
    payments: PaymentIntents[];
    paymentStatus: string;
    orderStatus: string;
    order_items: OrderItems[];
}

export abstract class IQuery {
    __typename?: 'IQuery';

    abstract user(id: string): Users | Promise<Users>;

    abstract getLikes(): Nullable<ProductVariations[]> | Promise<Nullable<ProductVariations[]>>;

    abstract getProducts(GetProductsInput: GetProductsInput): Nullable<Products[]> | Promise<Nullable<Products[]>>;

    abstract getProductById(id: string): Nullable<Products> | Promise<Nullable<Products>>;

    abstract getCategories(): Nullable<GetCategoriesResponse[]> | Promise<Nullable<GetCategoriesResponse[]>>;

    abstract getCartItems(): Nullable<CartItems[]> | Promise<Nullable<CartItems[]>>;

    abstract getOrders(GetOrdersInput: GetOrdersInput): Nullable<Orders[]> | Promise<Nullable<Orders[]>>;

    abstract getOrder(order_id: string): Nullable<Orders> | Promise<Nullable<Orders>>;
}

export abstract class IMutation {
    __typename?: 'IMutation';

    abstract toggleLike(product_variation_id: string): Nullable<LikesOfProducts> | Promise<Nullable<LikesOfProducts>>;

    abstract createProduct(CreateProductInput: CreateProductInput): Nullable<Products> | Promise<Nullable<Products>>;

    abstract updateProduct(UpdateProductInput: UpdateProductInput): Nullable<Products> | Promise<Nullable<Products>>;

    abstract deleteProduct(product_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract updateProductVariation(UpdateProductVariationInput: UpdateProductVariationInput): Nullable<ProductVariations> | Promise<Nullable<ProductVariations>>;

    abstract createProductVariation(CreateProductVariationInput: CreateProductVariationInput): Nullable<ProductVariations> | Promise<Nullable<ProductVariations>>;

    abstract deleteProductVariation(variation_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract createCategory(name: string): Nullable<Categories> | Promise<Nullable<Categories>>;

    abstract deleteCategory(id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract addToCart(variation_id: string, quantity: number): Nullable<CartItems> | Promise<Nullable<CartItems>>;

    abstract removeFromCart(product_variation_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract createOrder(): Nullable<Orders> | Promise<Nullable<Orders>>;
}

export type DateTime = any;
type Nullable<T> = T | null;
