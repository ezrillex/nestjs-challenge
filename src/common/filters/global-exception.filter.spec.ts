import { GlobalExceptionFilter } from './global-exception.filter';
import { ArgumentsHost, BadRequestException, HttpServer } from '@nestjs/common';
import { GraphQLError, GraphQLFormattedError } from 'graphql/error';

describe('Global Exception Filter', () => {
  let filter: GlobalExceptionFilter;
  let host: ArgumentsHost;
  let server: HttpServer;

  beforeEach(async () => {
    jest.useFakeTimers();
    jest.setSystemTime(new Date(2020, 12, 12, 12, 12, 12));

    server = {
      reply: jest.fn(),
    } as unknown as HttpServer;

    host = {
      getType: jest.fn(),
      switchToHttp: jest.fn().mockReturnValue({
        getRequest: jest.fn().mockReturnValue({
          url: '/test',
        }),
        getResponse: jest.fn().mockReturnValue({
          message: 'test response',
        }),
      }),
    } as unknown as ArgumentsHost;
    filter = new GlobalExceptionFilter(server);
  });

  it('should be defined', () => {
    expect(filter).toBeDefined();
  });

  it('should be defined', () => {
    expect(filter.catch).toBeDefined();
  });

  it('should handle an HTTP exception', async () => {
    const badRequest = new BadRequestException('Test message');

    const spy = jest.fn().mockResolvedValue(null);
    server.reply = spy;

    filter.catch(badRequest, host);

    expect(spy.mock.calls).toMatchSnapshot('http error reply');
  });

  it('should handle a GraphQL exception', async () => {
    const gqlError = new GraphQLError('Test error message');
    host.getType = jest.fn().mockReturnValue('graphql');

    server.reply = jest.fn().mockResolvedValue(null);

    filter.catch(gqlError, host);

    expect(filter.catch(gqlError, host)).toMatchSnapshot('gql error');
  });

  it('should handle a formatted GraphQL exception', async () => {
    const gqlException: GraphQLFormattedError = {
      message: `NotFoundException: error message for not found error`,
      extensions: {
        code: 404,
        timestamp: new Date().toISOString(),
        path: ['someTestPath'],
      },
    };

    host.getType = jest.fn().mockReturnValue('graphql');

    expect(filter.catch(gqlException, host)).toMatchSnapshot(
      'gql formatted exception',
    );
  });
});
