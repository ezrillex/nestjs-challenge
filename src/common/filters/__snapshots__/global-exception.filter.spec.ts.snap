// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global Exception Filter should handle a GraphQL exception: gql error 1`] = `[GraphQLError: Test error message]`;

exports[`Global Exception Filter should handle a formatted GraphQL exception: gql formatted exception 1`] = `
{
  "extensions": {
    "code": 404,
    "path": [
      "someTestPath",
    ],
    "timestamp": "2021-01-12T18:12:12.000Z",
  },
  "message": "NotFoundException: error message for not found error",
}
`;

exports[`Global Exception Filter should handle an HTTP exception: http error reply 1`] = `
[
  [
    {
      "message": "test response",
    },
    {
      "details": "Test message",
      "message": "Test message",
      "path": "/test",
      "status_code": 400,
      "timestamp": "2021-01-12T18:12:12.000Z",
    },
    400,
  ],
]
`;
